cluster-autoscaler/core/scaleup/orchestrator/executor.go

```diff
@@ -0,0 +1,240 @@
+		go func(info nodegroupset.ScaleUpInfo) {
+			defer wg.Done()
+			nodeInfo, ok := nodeInfos[info.Group.Id()]
+			if !ok {
+				klog.Errorf("ExecuteScaleUp: failed to get node info for node group %s", info.Group.Id())
+				return
+			}
+			if aErr := e.executeScaleUp(info, nodeInfo, availableGPUTypes, now); aErr != nil {
+				errResults <- errResult{err: aErr, info: &info}
+			}
+		}(scaleUpInfo)
+	}
+	wg.Wait()
+	close(errResults)
+	var results []errResult
+	for err := range errResults {
+		results = append(results, err)
+	}
+	if len(results) > 0 {
+		failedNodeGroups := make([]cloudprovider.NodeGroup, len(results))
+		scaleUpErrors := make([]errors.AutoscalerError, len(results))
+		for i, result := range results {
+			failedNodeGroups[i] = result.info.Group
+			scaleUpErrors[i] = result.err
+		}
+		return combineConcurrentScaleUpErrors(scaleUpErrors), failedNodeGroups
+	}
+	return nil, nil
+}
+
+func (e *scaleUpExecutor) executeScaleUp(
+	info nodegroupset.ScaleUpInfo,
+	nodeInfo *schedulerframework.NodeInfo,
+	availableGPUTypes map[string]struct{},
+	now time.Time,
+) errors.AutoscalerError {
+	gpuConfig := e.autoscalingContext.CloudProvider.GetNodeGpuConfig(nodeInfo.Node())
+	gpuResourceName, gpuType := gpu.GetGpuInfoForMetrics(gpuConfig, availableGPUTypes, nodeInfo.Node(), nil)
+	klog.V(0).Infof("Scale-up: setting group %s size to %d", info.Group.Id(), info.NewSize)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: setting group %s size to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	increase := info.NewSize - info.CurrentSize
+	if err := info.Group.IncreaseSize(increase); err != nil {
+		e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeWarning, "FailedToScaleUpGroup", "Scale-up failed for group %s: %v", info.Group.Id(), err)
+		aerr := errors.ToAutoscalerError(errors.CloudProviderError, err).AddPrefix("failed to increase node group size: ")
+		e.clusterStateRegistry.RegisterFailedScaleUp(info.Group, metrics.FailedScaleUpReason(string(aerr.Type())), now)
+		return aerr
+	}
+	e.clusterStateRegistry.RegisterOrUpdateScaleUp(
+		info.Group,
+		increase,
+		time.Now())
+	metrics.RegisterScaleUp(increase, gpuResourceName, gpuType)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: group %s size set to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	return nil
+}
+
+func combineConcurrentScaleUpErrors(errs []errors.AutoscalerError) errors.AutoscalerError {
+	if len(errs) == 0 {
+		return nil
+	}
+	if len(errs) == 1 {
+		return errs[0]
+	}
+	// sort to stabilize the results and easier log aggregation
+	sort.Slice(errs, func(i, j int) bool {
+		errA := errs[i]
+		errB := errs[j]
+		if errA.Type() == errB.Type() {
+			return errs[i].Error() < errs[j].Error()
+		}
+		return errA.Type() < errB.Type()
+	})
+	firstErrMsg := errs[0].Error()
```

Comment:
Why can't we make two `map[string]bool` one having types and second messages?
This way `len(typesMap) > 1` should be equivalent to  `singleType` variable

Reply:
Yes, that's right but we would simply replace two booleans with two maps, making it a bit less optimal. Is it a readability issue?